<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro"  name="robot">

    <xacro:include filename="inertial_macros.xacro" />

    <!-- This first link called "world" is empty -->
    <link name="world"></link>

    <joint name="base_joint" type="fixed">
        <origin xyz="0.5 0.5 0" rpy="0 0 0"/>
        <parent link="world"/>
        <child link="base_link"/>        
    </joint>



    <link name="base_link">
        <visual>
            <origin xyz="0 0 0.05" rpy="0 0 0"/>
            <geometry>
                <box size="1 1 0.1" />
            </geometry>
            <material name="green">
                <color rgba="0.2 1 0.2 1"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 0.05" rpy="0 0 0"/>
            <geometry>
                <box size="1 1 0.1" />
            </geometry>
        </collision>
        <inertial>
            <origin xyz="0 0 0.05" rpy="0 0 0"/>
            <mass value="12" />
            <inertia ixx="2.26" ixy="0.0" ixz="0.0" iyy="6.26" iyz="0.0" izz="8.5" />
        </inertial>
    </link>



    <joint name="joint_1" type="revolute">
        <origin xyz="0 0 0.1" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="link_1"/>
        <axis xyz="0 0 1"/>
        <limit lower="0" upper="5.585" velocity="1" effort="100"/> 
        <dynamics damping="1.0" friction="1.0"/>
    </joint>



    <xacro:property name="link_1_length" value="${0.59817}" />
    <xacro:property name="link_1_radius" value="0.07493" />
    <link name="link_1">
    <visual>
        <origin xyz="0 0 ${link_1_length/2}" rpy="0 0 0"/>
        <geometry>
            <!-- <cylinder length="${link_1_length}" radius="${link_1_radius}" /> -->
                <mesh filename="package://arm_bot/meshes/base_link.STL" scale="1 1 1"/>
        </geometry>
        <material name="blue" />
    </visual>
    <collision>
        <origin xyz="0 0 ${link_1_length/2}" rpy="0 0 0"/>
        <geometry>
            <cylinder length="${link_1_length}" radius="${link_1_radius}" />
        </geometry>
    </collision>
    <xacro:inertial_cylinder mass="0.5" length="${link_1_length}" radius="${link_1_radius}">
        <origin xyz="0 0 ${link_1_length/2}" rpy="0 0 0"/>
    </xacro:inertial_cylinder>
    </link>


    
    <!-- Add properties for shoulder -->
    <xacro:property name="shoulder_length" value="${0.1686 + link_1_radius}" />
    <xacro:property name="shoulder_radius" value="${link_1_radius}" />
    
    <joint name="joint_1_shoulder" type="fixed">
        <origin xyz="0 0 ${link_1_length + shoulder_radius}" rpy="0 0 0"/>
        <parent link="link_1"/>
        <child link="shoulder"/>
    </joint>



    <link name="shoulder">
        <visual>
            <origin xyz="0 ${shoulder_length/2 - shoulder_radius} 0" rpy="0 0 0"/>
            <geometry>
                <!-- <cylinder length="${shoulder_length}" radius="${shoulder_radius}" /> -->
                <mesh filename="package://arm_bot/meshes/link_1.STL" scale="1 1 1"/>
            </geometry>
            <material name="blue" />
        </visual>
        <collision>
            <origin xyz="0 ${shoulder_length/2 - shoulder_radius} 0" rpy="0 0 0"/>
            <geometry>
                <cylinder length="${shoulder_length}" radius="${shoulder_radius}" />
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.5" length="${shoulder_length}" radius="${shoulder_radius}">
            <origin xyz="0 ${shoulder_length/2 - shoulder_radius} 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>



    <xacro:property name="link_2_length" value="${0.735}" />
    <xacro:property name="link_2_radius" value="0.065278" />


    <joint name="joint_2" type="revolute">
        <origin xyz="0 ${shoulder_length} 0" rpy="0 0 0"/>
        <parent link="shoulder"/>
        <child link="link_2"/>
        <axis xyz="0 -1 0"/>
        <limit lower="-1" upper="3.363" velocity="1" effort="100"/> 
        <dynamics damping="1.0" friction="1.0"/>
    </joint>



    <link name="link_2">
        <visual>
            <origin xyz="${link_2_length/2 - 0.226} 0 0" rpy="0 0 0"/>
            <geometry>
                <!-- <cylinder length="${link_2_length}" radius="${link_2_radius}" /> -->
                <mesh filename="package://arm_bot/meshes/link_2.STL" scale="1 1 1"/>
            </geometry>
            <material name="orange" />
        </visual>
        <collision>
            <origin xyz="${link_2_length/2 - 0.226} 0 0" rpy="0 0 0"/>
            <geometry>
                <cylinder length="${link_2_length}" radius="${link_2_radius}" />
                <!-- <mesh filename="package://arm_bot/meshes/puma_link3.stl" scale="0.01 0.01 0.01"/> -->
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="1.0" length="${link_2_length}" radius="${link_2_radius}">
            <origin xyz="${link_2_length/2 - 0.226} 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>


    <xacro:property name="link_3_length" value="${0.4318 + 0.13208}" />
    <xacro:property name="link_3_radius" value="0.028702" />


    <joint name="joint_3" type="revolute">
        <origin xyz="${link_2_length - 0.226 - link_3_radius} -${link_2_radius + link_3_radius} 0" rpy="0 0 0"/>
        <parent link="link_2"/>
        <child link="link_3"/>
        <axis xyz="0 -1 0"/>
        <limit lower="-0.712" upper="4" velocity="1" effort="100"/> 
        <dynamics damping="1.0" friction="1.0"/>
    </joint>



    <link name="link_3">
        <visual>
            <origin xyz="0 0 ${link_3_length/2 - 0.4318}" rpy="0 0 0"/>
            <geometry>
                <!-- <cylinder length="${link_3_length}" radius="${link_3_radius}" /> -->
                <mesh filename="package://arm_bot/meshes/link_3.STL" scale="1 1 1"/>
            </geometry>
            <material name="yellow" />
        </visual>
        <collision>
            <origin xyz="0 0 ${link_3_length/2 - 0.4318}" rpy="0 0 0"/>
            <geometry>
                <cylinder length="${link_3_length}" radius="${link_3_radius}" />
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.5" length="${link_3_length}" radius="${link_3_radius}">
            <origin xyz="0 0 ${link_3_length/2 - 0.4318}" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

</robot>